recitation_3/structs-unions.md

```diff
@@ -0,0 +1,281 @@
+
+In the above example, the largest type is `char *`. Therefore, an extra 4 bytes of space are padded after `int len`. `sizeof(struct MyString)` then would evaluate to 16.
+
+A struct may pad differently depending on the order of elements. A poorly ordered struct may look something like this:
+
+```c
+struct MyStuff {
+    char *str1;
+    int s = len1;
+    char *str2;
+    int s len2;
+};
+
+Here, the two `char *`s are the largest type with 8 bytes. :
+
+```
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ | .str1                         |  .len1        |  (padding)    | .str2                         |  .len2        | (padding)     |
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ |-------------------------------------------------------- struct MyStuff -------------------------------------------------------|
+```
+
+However, if we are to reorganize the struct like so:
+
+```c
+struct MyStuff2 {
+    char *str1;
+    char *str2;
+    int len1;
+    int len2;
+};
+```
+The layout would look like this:
+
+```
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ | .str1                         | .str2                         |  .len1        | .len2         |
+ +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ |------------------------------------ struct MyStuff2 ---------------------- -------------------|
+```
+
+Try to guess what the size of each struct will be below:
+
+```c
+#include <stdio.h>
+
+struct a {
+    char *x;
+    int y;
+    char z;
+};
+
+struct b {
+    int x;
+    char *y;
+    char z;
+};
+
+struct c {
+    char *p;   
+    char c;
+    short x;
+};
+
+struct d {
+    char w;
+    int x;
+    short y;
+    char z;
+};
+
+
+int main() {
+    printf("%lu\n", sizeof(struct a));
+    printf("%lu\n", sizeof(struct b));
+    printf("%lu\n", sizeof(struct c));
+    printf("%lu\n", sizeof(struct d));
+}
+```
+
+### Pointers to structs
+
+Since structs may get very large in size (e.g., struct Buffer), it can be more
+efficient to refer to them via pointers rather than copy them around by value:
+
+```c
+void inefficient_arg(struct Buffer buf) { ... }
+void efficient_arg(struct Buffer *buf) { ... }
+```
+
+Thus, a common pattern when using a pointer to a struct is to first dereference
+it, and then access the field:
+
+```c
+(*b).len
+```
+
+C provides syntactic sugar to make this look nicer:
+
+```c
+b->len
+```
+
+This notation should be read as, "dereference b, and access the len field."
+Keep in mind that this notation still performs a dereference; if b is a NULL
+pointer, this expression will segfault.
```

Comment:
```suggestion
Keep in mind that this notation still performs a dereference. If b is a NULL pointer, this expression will segfault.
```
Also idk about this, this is undefined behavior, it will segfault in most cases but this is not in the satndard